lib/s3middleware/objectRetention.js

```diff
@@ -0,0 +1,125 @@
+const { parseString } = require('xml2js');
+
+const errors = require('../errors');
+
+/*
+    Format of xml request:
+    <Retention>
+      <Mode>COMPLIANCE|GOVERNANCE</Mode>
+      <RetainUntilDate>2020-05-20T04:58:45.413000Z</RetainUntilDate>
+    </Retention>
+*/
+
+function validateMode(mode) {
+    const validMode = {};
+    const expectedModes = ['GOVERNANCE', 'COMPLIANCE'];
+    if (!mode || !mode[0]) {
+        validMode.error = errors.MalformedXML.customizeDescription(
+            'request xml does not contain Mode');
+        return validMode;
+    }
+    if (mode.length > 1) {
+        validMode.error = errors.MalformedXML.customizeDescription(
+            'request xml contains more than one Mode');
+        return validMode;
+    }
+    if (!expectedModes.includes(mode[0])) {
+        validMode.error = errors.MalformedXML.customizeDescription(
+            'Mode request xml must be one of "GOVERNANCE", "COMPLIANCE"');
+        return validMode;
+    }
+    validMode.mode = mode[0];
+    return validMode;
+}
+
+function validateRetainDate(retainDate) {
+    const validRetainDate = {};
+    if (!retainDate || !retainDate[0]) {
+        validRetainDate.error = errors.MalformedXML.customizeDescription(
+            'request xml does not contain RetainUntilDate');
+        return validRetainDate;
+    }
+    if (!/\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}.\d{3}Z/.test(retainDate[0])) {
```

Comment:
Although this looks right, I can't help but think javascript has a built-in date checker. Or something close. Might need a plugin but even ISO-8601 I believe has a few variants or optional components that we would need to check for.

Reply:
hey @dora-korpar you marked this as resolved but I don't see any change or response mentioning why the code should be kept.

Reply:
If there's no standard function to check it, it could also likely be a reusable constant regexp.